src/code-components/RichTextArea/Editor.tsx

```diff
@@ -0,0 +1,148 @@
+import React, { forwardRef, useEffect, useLayoutEffect, useRef } from "react";
+import "quill/dist/quill.snow.css";
+import { Delta as DeltaType, EmitterSource, Range } from "quill/core";
+import { ToolbarConfigs } from "./formatDefaultToolbarConfigs";
+import Quill from "quill";
+
+interface EditorProps {
+  defaultValue?: DeltaType | string;
+  toolbarConfigs?: ToolbarConfigs | false;
+  onTextChange?: (content: string, source: EmitterSource) => void;
+  onSelectionChange?: (range: Range, source: EmitterSource) => void;
+  onBlur?: (range: Range, source: EmitterSource) => void;
+  onFocus?: (range: Range, source: EmitterSource) => void;
+  onKeyDown?: (event: KeyboardEvent) => void;
+  onKeyUp?: (event: KeyboardEvent) => void;
+  placeholder?: string;
+  readOnly?: boolean;
+  wrapperClassName?: string;
+  ariaLabel?: string;
+  ariaLabeledby?: string;
+  role?: string;
+}
+
+export const Editor = forwardRef<Quill | null, EditorProps>(
+  (
+    {
+      defaultValue,
+      toolbarConfigs,
+      onTextChange,
+      onSelectionChange,
+      onBlur,
+      onFocus,
+      onKeyDown,
+      onKeyUp,
+      placeholder,
+      readOnly,
+      wrapperClassName,
+      ariaLabel,
+      ariaLabeledby,
+      role,
+    },
+    ref,
+  ) => {
+    const containerRef = useRef<HTMLDivElement | null>(null);
+    const defaultValueRef = useRef(defaultValue);
+    const onTextChangeRef = useRef(onTextChange);
+    const onSelectionChangeRef = useRef(onSelectionChange);
+    const onBlurRef = useRef(onBlur);
+    const onFocusRef = useRef(onFocus);
+    const onKeyDownRef = useRef(onKeyDown);
+    const onKeyUpRef = useRef(onKeyUp);
+
+    useLayoutEffect(() => {
+      onTextChangeRef.current = onTextChange;
+      onSelectionChangeRef.current = onSelectionChange;
+    });
```

Comment:
why do we do that in uselayout effect, rather than always?

also, what about other event handlers in refs - why don't we keep them up-to-date?

```suggestion
    onTextChangeRef.current = onTextChange;
    onSelectionChangeRef.current = onSelectionChange;
```